tests/python/gstaichi/lang/fast_caching/test_src_hasher.py

```diff
@@ -0,0 +1,145 @@
+    cache_key_base = src_hasher.create_cache_key(kernel_info, [])
+
+    ti_init_same_arch()
+    kernel_info, _src = get_source_info_and_src(f1.fn)
+    cache_key_same = src_hasher.create_cache_key(kernel_info, [])
+
+    ti_init_same_arch(options={"offline_cache_max_size_of_files": 123})
+    kernel_info, _src = get_source_info_and_src(f1.fn)
+    cache_key_diff = src_hasher.create_cache_key(kernel_info, [])
+
+    assert cache_key_base == cache_key_same
+    assert cache_key_same != cache_key_diff
+
+
+@test_utils.test()
+def test_src_hasher_create_cache_key_vary_fn(monkeypatch) -> None:
+    test_files_path = "tests/python/gstaichi/lang/fast_caching/test_files"
+    monkeypatch.syspath_prepend(test_files_path)
+
+    def get_cache_key(name: str) -> _wrap_inspect.FunctionSourceInfo:
+        mod = importlib.import_module(name)
+        info, _src = _wrap_inspect.get_source_info_and_src(mod.f1.fn)
+        cache_key = src_hasher.create_cache_key(info, [])
+        return cache_key
+
+    key_base = get_cache_key("f1_base")
+    print(key_base)
+    key_same = get_cache_key("f1_same")
+    print(key_same)
+    key_diff = get_cache_key("f1_diff")
+    print(key_diff)
+
+    assert key_base is not None
+    assert key_base != ""
+    assert key_base == key_same
+    assert key_base != key_diff
+
+
+@test_utils.test()
+def test_src_hasher_validate_hashed_function_infos(monkeypatch, tmp_path: pathlib.Path) -> None:
+    test_files_path = pathlib.Path("tests/python/gstaichi/lang/fast_caching/test_files")
+    monkeypatch.syspath_prepend(str(tmp_path))
+
+    def setup_folder(filename: str) -> None:
+        shutil.copy2(test_files_path / filename, tmp_path / "child_diff.py")
+
+    setup_folder("child_diff_base.py")
+    mod = importlib.import_module("child_diff")
+
+    def get_fileinfos(functions: list[Callable]) -> list[_wrap_inspect.FunctionSourceInfo]:
+        fileinfos = []
+        for f in functions:
+            file_info, _src = _wrap_inspect.get_source_info_and_src(f)
+            fileinfos.append(file_info)
+        return fileinfos
+
+    fileinfos = get_fileinfos([mod.f1.fn, mod.f2.fn])
+    hashed_fileinfos = function_hasher.hash_functions(fileinfos)
+    assert function_hasher.validate_hashed_function_infos(hashed_fileinfos)
+
+    setup_folder("child_diff_same.py")
+    assert function_hasher.validate_hashed_function_infos(hashed_fileinfos)
+
+    setup_folder("child_diff_diff.py")
+    assert not function_hasher.validate_hashed_function_infos(hashed_fileinfos)
+
+    setup_folder("child_diff_same.py")
+    assert function_hasher.validate_hashed_function_infos(hashed_fileinfos)
+
+
+@test_utils.test()
+def test_src_hasher_store_validate(monkeypatch: pytest.MonkeyPatch, tmp_path: pathlib.Path) -> None:
+    test_files_path = pathlib.Path("tests/python/gstaichi/lang/fast_caching/test_files")
+
+    tmp_path = tmp_path / test_src_hasher_store_validate.__name__
+    tmp_path.mkdir(exist_ok=True)
```

Comment:
`tmp_path` is already specific to every single test. What is your point?

Reply:
The imports are not. But actually this didnt fix the imports. But I found a way to fix the imports now.